### DISCUSSION TITLE
Отсутствующие ключи в ответе внешнего API

### DISCUSSION THREAD

**Discussion by andlev7** [totumonline/totum-ru-issues-and-discussions#174] [Есть идея, а что если...]
Появилась задача распарсить результаты JSON полученных в результате запроса по API к стороннему сервису. Спасибо Алексею, что сделал готовый скрипт, размещаю здесь, чтобы могли использовать если у вас похожая задача. Исходный JSON, 

![image](https://user-images.githubusercontent.com/52709157/182152596-beac29da-32a9-429b-8213-d3906ebd1fae.png)

**Код действия Тотум:**

```
=: insertList(table: 'api_test'; field: 'name' = $data_from_api[[domain]]; field: 'keywords' = $data_from_api[[keywords]]; field: 'traffic' = $data_from_api[[traff]]; field: 'visible' = $data_from_api[[visible]])

~data_from_api: $json_extr[result][data]

json_extr: "ЗДЕСЬ ТО, ЧТО ПРИШЛО ПО API ЗАПРОСУ"

// insertList даст вам одно действие записи в БД вместо 10 при использовании while
// [[]] двойные скобки — это взять колонку из rowlist в виде list
// ~ это фиксация — те data_from_api будет рассчитано один раз, если без ~ то 4 раза (по количество обращения) и не факт, что порядок будет одинаковый
```

**При этом возникла ошибка Ключ `domain` не обнаружен в одном из элементов массива так-как так отдает запрашиваемый сервер** 

Решается двумя путями добавить строку в коде вариант 1 и 2:

**Вариант 1:**

```
// фильтровать по непустому значению в `domain` и пропустить те строки, в которых его нет при помощи skip: true

data_from_api: listFilter(list: $json_extr[result][data]; key: "domain" != ""; skip: true)

~json_extr: "ЗДЕСЬ ТО, ЧТО ПРИШЛО ПО API ЗАПРОСУ"

```

**Вариант 2:**

```
// Перебрать все строки и заменить отсутствующий ключ на пустое значение

data_from_api: listReplace(list: $json_extr[result][data]; action: "domain" = $domain; value: $#row)

domain: $#row[domain]

~json_extr: "ЗДЕСЬ ТО, ЧТО ПРИШЛО ПО API ЗАПРОСУ"
```

**Comment by ttm-alexeypolunin** [comment]
Я переместил в **Есть идея** так как хорошо бы сделать функцию, которая будет автоматически дополнять отсутствующие ключи. В ответах API такое встречается, что если значение ключа пустое, то ключ не передается. Это экономия на спичках, но разработчиков сторонних API не переубедишь :)
